### PR TITLE
Prometheus Node Exporter

### PR DESCRIPTION
**Pull request by kfox1111**
Adds a chart for managing the Prometheus Node Exporter as a
Daemonset.

### CODE REVIEW

**Comment by sstarcher** [comment]
Do we want a separate node exporter chart now that the prometheus chart will have the node exporter?

https://github.com/kubernetes/charts/pull/627

Or should #627 be updated to reference this as a dependency.

**Comment by mgoodness** [comment]
@sstarcher The current plan is to get #627 merged, followed by all of the stand-alone components, then refactor using dependencies. It's just taken far too long to accomplish that first step due to external factors.

**Comment by sstarcher** [comment]
Sounds awesome, keep up the good work

**Comment by kfox1111** [comment]
Having it separate allows the user to configure versions of the node exporter differently depending on the needs of various pools of hardware.

If the long term goal is to split it out anyway, why not just drop support for the node exporter in the new Prometheus chart for now, and we can work on getting it fixed up here?

That does raise another discussion point. There are kind of two use cases for charts that we've hit for Kolla-Kubernetes. Charts to do one thing, such as a node exporter or a Prometheus server, and there are charts that are for deploying more complex systems, such as a "prometheus + common exporters" group of charts.
How are we going to handle the group use case in kubernetes/charts?

**Comment by mgoodness** [comment]
> If the long term goal is to split it out anyway, why not just drop support for the node exporter in the new Prometheus chart for now, and we can work on getting it fixed up here?

#627 was already at the end of its review, while this one (and the other components) have yet to begin. I wanted to get something out there for people to use.

> How are we going to handle the group use case in kubernetes/charts?

I expect to see most multi-component charts broken up into individual ones and new "umbrella" versions created that allow specific dependencies to be enabled or disabled by the end-user. Ultimately more charts, but smaller and more flexible ones.

**Comment by kfox1111** [comment]
Sounds good. If there are no more revisions of the other chart, and its to merge quickly, then yeah, we can just let it go through and then in a follow on pr, pull it back out. Though arguably that may break compatibility. Is there a policy there about backwards compatibility?

yeah. umbrella charts. thats a good way of putting it. So, would 'prometheus' become an an umbrella chart under stable, would just the server bits become prometheus-server? or would prometheus be just the server bits and an umbrella chart somewhere else would also be called prometheus or maybe prometheus-all or something?

**Comment by mgoodness** [comment]
> Is there a policy there about backwards compatibility?

We don't have a policy beyond "use SemVer correctly". I'd personally (and generally) like to see backwards compatibility retained unless there's a compelling reason not to.

> umbrella charts

I _think_ the `prometheus` chart wants to define the `server` component, with alertmanager, node exporter, cloudwatch exporter, kube-state-metrics, etc. as optional dependencies. The idea being anyone who wants to install just prometheus can use that chart with everything else disabled. But if they want the kitchen sink, they can enable it all.

**Comment by kfox1111** [comment]
So, prometheus would be the umbrella chart sort of and the primary chart for the server? Which way would it default? everything else disabled or enabled?

**Comment by mgoodness** [comment]
Turning everything else off seems like the best (only?) way to avoid a fight over what should be enabled and what shouldn't. Also, Prometheus configuration is so widely variable that I doubt anyone deploys it without some customization anyway.

**Comment by kfox1111** [comment]
the other route is to have 2 charts. one for just prometheus server bits, and one for everything. Then if all you want is prometheus, you just use the server only version. Lets you more easily embed it in other charts maybe?

**Comment by prydonius** [comment]
ping @mgoodness

**Comment by kfox1111** [comment]
So, whats the path forward here?

**Comment by mgoodness** [comment]
@kfox1111 I'd like to see the PR updated to use the existing `stable/prometheus/templates/node-exporter-*` manifests as the basis for this chart. That'll make it easier for us to pull those manifests out and drop this in as an (optional) dependency.

**Comment by kfox1111** [comment]
@mgoodness 
so, I'm hitting an issue with that which I was a afraid of  before letting it merge as it was.

The whole node-exporter thing could be lifted out as its own chart, and then using local path deps pulled back into the prometheus chart with a conditional include to get back to the existing functionality.

But there's a problem. Convention. the package should be named node-exporter by convention but the section name established in the values file is nodeExporter by value convention. The only renaming thing I could find was alias but doesn't allow you to rename, just add more things.

I'm also struggling with getting my prometheus systems that were installed with the previous version of the chart to the new version. a lot has changed. :/

So getting it to the point I can contribute is a bit hard now. I'm working on it though.

should we just rename the section node-exporter, or should we try and land a feature in helm that would allow renaming the section of values.yaml the subchart reads? @technosophos (you might be interested in this too) This may become a more common issue as subcharts are created from primary charts when following established naming conventions.

**Comment by unguiculus** [comment]
ping @mgoodness

**Comment by kfox1111** [comment]
I've split out the bits that are in the Prometheus chart out verbatim:
https://github.com/kubernetes/charts/pull/1886

Once that lands we can merge the couple of extra tweaks I've done in this pr?

File: stable/node-exporter/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name: node-exporter
+version: 0.13.0
+description: Prometheus node exporter module
+keywords:
+- prometheus
+- node
+maintainers:
+  - name: Kevin Fox
```
Comment (unguiculus) [inline stable/node-exporter/Chart.yaml:8]: Use Github username

File: stable/node-exporter/templates/daemonset.yaml [hunk 2]
```diff
@@ -0,0 +1,74 @@
+apiVersion: extensions/v1beta1
```
Comment (unguiculus) [inline stable/node-exporter/templates/daemonset.yaml:1]: `apps/v1beta2`

File: stable/node-exporter/templates/daemonset.yaml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+      annotations:
+        "prometheus.io/port": "9100"
+        "prometheus.io/scrape": "true"
+    spec:
+      hostNetwork: True
+      nodeSelector:
+{{ toYaml .Values.nodeSelector | indent 10 }}
+      containers:
+      - name: {{ template "fullname" . }}
+        image: "{{default "quay.io/prometheus/node-exporter" .Values.image}}:{{default "v0.13.0" .Values.imageTag}}"
```
Comment (unguiculus) [inline stable/node-exporter/templates/daemonset.yaml:25]: Don't use `default` function if not necessary. Defaults should go into `values.yaml`.

File: stable/node-exporter/templates/_helpers.tpl [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{{/* vim: set filetype=mustache: */}}
```
Comment (unguiculus) [inline stable/node-exporter/templates/_helpers.tpl:1]: Use `_helpers.tpl` you get with `helm create`

File: stable/node-exporter/values.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# Default values for ceph-exporter.
+# This is a YAML-formatted file.
+# Declare name/value pairs to be passed into your templates.
+# name: value
+
+Name: node-exporter
```
Comment (unguiculus) [inline stable/node-exporter/values.yaml:6]: Remove this.

File: stable/node-exporter/templates/daemonset.yaml [hunk 6]
```diff
@@ -0,0 +1,74 @@
+apiVersion: extensions/v1beta1
+kind: DaemonSet
+metadata:
+  name: {{template "fullname" .}}
```
Comment (unguiculus) [inline stable/node-exporter/templates/daemonset.yaml:4]: Use consistent formatting: Space after `{{` and  before `}}`.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for the PR. Please read our review guidelines and follow best practices.

https://github.com/unguiculus/charts/blob/master/REVIEW_GUIDELINES.md

**Comment by unguiculus** [comment]
Just realized there's also #1886. It this one obsolete now?

**Comment by kfox1111** [comment]
Yeah, #1886 is the one that was pulled back out of the merged code base. its closer to trunk so probably the better one to adopt.

**Pull request closed without merging**